packages/aws-cdk/lib/api/hotswap/appsync-mapping-templates.ts

```diff
@@ -49,25 +50,76 @@ export async function isHotswappableAppSyncChange(
 
         const sdkProperties: { [name: string]: any } = {
           ...change.oldValue.Properties,
+          Definition: change.newValue.Properties?.Definition,
+          DefinitionS3Location: change.newValue.Properties?.DefinitionS3Location,
           requestMappingTemplate: change.newValue.Properties?.RequestMappingTemplate,
+          requestMappingTemplateS3Location: change.newValue.Properties?.RequestMappingTemplateS3Location,
           responseMappingTemplate: change.newValue.Properties?.ResponseMappingTemplate,
+          responseMappingTemplateS3Location: change.newValue.Properties?.ResponseMappingTemplateS3Location,
         };
         const evaluatedResourceProperties = await evaluateCfnTemplate.evaluateCfnExpression(sdkProperties);
         const sdkRequestObject = transformObjectKeys(evaluatedResourceProperties, lowerCaseFirstCharacter);
 
+        // resolve s3 location files as SDK doesn't take in s3 location but inline code
+        if (sdkRequestObject.requestMappingTemplateS3Location) {
+          sdkRequestObject.requestMappingTemplate = (await fetchFileFromS3(sdkRequestObject.requestMappingTemplateS3Location, sdk))?.toString('utf8');
+          delete sdkRequestObject.requestMappingTemplateS3Location;
+        }
+        if (sdkRequestObject.responseMappingTemplateS3Location) {
+          sdkRequestObject.responseMappingTemplate = (await fetchFileFromS3(sdkRequestObject.responseMappingTemplateS3Location, sdk))?.toString('utf8');
+          delete sdkRequestObject.responseMappingTemplateS3Location;
+        }
+        if (sdkRequestObject.definitionS3Location) {
+          sdkRequestObject.definition = await fetchFileFromS3(sdkRequestObject.definitionS3Location, sdk);
+          delete sdkRequestObject.definitionS3Location;
+        }
+
         if (isResolver) {
           await sdk.appsync().updateResolver(sdkRequestObject).promise();
-        } else {
+        } else if (isFunction) {
+
           const { functions } = await sdk.appsync().listFunctions({ apiId: sdkRequestObject.apiId }).promise();
           const { functionId } = functions?.find(fn => fn.name === physicalName) ?? {};
-          await sdk.appsync().updateFunction({
-            ...sdkRequestObject,
-            functionId: functionId!,
-          }).promise();
+          await simpleRetry(
+            () => sdk.appsync().updateFunction({ ...sdkRequestObject, functionId: functionId! }).promise(),
+            3,
+            'ConcurrentModificationException');
+        } else {
+          let schemaCreationResponse: GetSchemaCreationStatusResponse = await sdk.appsync().startSchemaCreation(sdkRequestObject).promise();
+          while (schemaCreationResponse.status && ['PROCESSING', 'DELETING'].some(status => status === schemaCreationResponse.status)) {
+            await new Promise(resolve => setTimeout(resolve, 1000)); // poll every second
```

Comment:
can you make a little helper for this, similar to: https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk/lib/util/archive.ts#L84

Reply:
Good idea, updated.